waffle_hub/core/hpo/adapter/optuna/optuna_hpo.py

```diff
@@ -0,0 +1,524 @@
+        if not self.hpo_artifacts_dir.exists():
+            io.make_directory(self.hpo_artifacts_dir)
+        for visualize in [
+            "plot_param_importances",
+            "plot_contour",
+            "plot_parallel_coordinate",
+            "plot_slice",
+            "plot_optimization_history",
+        ]:
+            self.draw_hpo_plot(visualize)
+
+    def _get_search_space(self, trial: optuna.trial.Trial, search_space: dict, is_hub: bool) -> dict:
+        advance_params = {}
+        params = {}
+        for k, v in search_space.items():
+            if k == "advance_params":
+                for k2, v2 in v.items():
+                    choice_value = ChoiceMethod(
+                        choice_name=k2,
+                        method_name=v2["method"],
+                        choices=v2["search_space"],
+                        **v2["kwargs"],
+                    )(trial)
+                    advance_params[k2] = choice_value
+            else:
+                choice_value = ChoiceMethod(
+                    choice_name=k, method_name=v["method"], choices=v["search_space"], **v["kwargs"]
+                )(trial)
+                params[k] = choice_value
+
+        if is_hub:
+            params.update(advance_params)
+
+        return params
+
+    # def _get_best_params(self, best_params: dict):
+    #     advance_params = {}
+    #     params = {}
+    #     for k, v in self.search_space.items():
+    #         if k == "advance_params":
+    #             for k2 in v.keys():
+    #                 advance_params[k2] = best_params[k2]
+    #         else:
+    #             params[k] = best_params[k]
+    #     params.update({"advance_params" : advance_params})
+    #     return params
+
+    def _get_best_params(self, best_params: dict):
+        params = {}
+        for k, v in self.search_space.items():
+            if k == "advance_params":
+                params[k] = {k2: best_params[k2] for k2 in v.keys()}
+            else:
+                params[k] = best_params[k]
+        return params
+
+    def optimize(self, objective: Callable, **kwargs) -> None:
+        def objective_wrapper(trial: optuna.trial.Trial) -> float:
+            search_space_values = self._get_search_space(trial, self.search_space, self.is_hub)
+
+            if self.is_hub:
+                search_space_values.update({"metric": self.metric, "trial": trial})
+
+            kwargs.update(search_space_values)
+            return objective(**kwargs)
+
+        self._study.optimize(objective_wrapper, n_trials=self.n_trials)
+
+    def load_hpo(
+        self,
+        root_dir: str,
+        study_name: str,
+    ):
+        self._study = optuna.load_study(
+            study_name=self._study_name, storage=f"sqlite:///{root_dir}/{study_name}/{study_name}.db"
+        )
+        return self._study
+
+    def run_hpo(
+        self,
+        objective: Callable,
+        visualize_hpo: bool = True,
+        **kwargs,
+    ) -> HPOResult:
```

Comment:
Why is the `run_hpo` part of `hub.py` different from the parameters here?
Please explicitly define the required parameters.

And, it needs to be modified so that it can operate on the background to match the `hold` option.

Reply:
Since optuna hpo's run_hpo method can receive not only the hub objective function but also other objective functions as input, I decided that there was no reason to explicitly specify the required parameters used in the hub's hpo method in optuna hpo's run_hpo.